File: 2nd round/tax-saver/tax-saver.cpp
#include <iostream>
#include<bits/stdc++.h>
using namespace std;

int main() {
    // your code goes here
    int i,j,k,n,m,x,y,z,q,u,v;
    int ans;
    cin>>n>>q;
    int a[n];
    cin>>a[0];
    // x=a[0];
    y=a[0];
    for(i=1;i<n;i++)
    {
    cin>>a[i];
    //x=min(x,a[i]);
    y=max(y,a[i]);
    }
    //u=0;
    v=y;
    sort(a,a+n);
    int b[n];
    b[n-1]=a[n-1];
    for(i=n-2;i>=0;i--)
    b[i]=a[i]+b[i+1];
    // x=1;
    // y=1000000;
    while(q--)
    {
        cin>>m;
        x=0;
        y=v;
    while(x<=y)
    {
        z=(x+y)/2;
        i=upper_bound(a,a+n,z)-a;
        if(i<n)
        {
        //j=i*z;
        j=b[i];
        j-=(n-i)*z;
        if(j>=m)
        {
            x=z+1;
            ans=z;
        }
            else
            {
            y=z-1;    
            }
        }
        else
        y=z-1;
    }
    cout<<ans<<" ";
    }
    cout<<"\n";
    return 0;
}

File: 2nd round/tax-saver/tax-saver.java
import java.util.*;
import java.util.*;
public class Solution {
    static int upper_bound(int array[], int key)
    {
        int index = Arrays.binarySearch(array, key);
        if (index < 0) {
            return Math.abs(index) - 1;
        }
        else {
            while (index > 0) {
                if (array[index - 1] == key)
                    index--;
                else
                    return index;
            }
 
            return index;
        }
    }  
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n=sc.nextInt();
        int q=sc.nextInt();
        int a[] = new int[n];
        int x=Integer.MAX_VALUE;
        int y=Integer.MIN_VALUE;

        for(int i=0;i<n;i++){
            a[i]=sc.nextInt();
            x=Math.min(x,a[i]);
            y=Math.max(y,a[i]);
        }
        int u=x;
        int v=y;
        Arrays.sort(a);
        int b[]=new int[n];
        b[n-1]=a[n-1];
        for(int i=n-2;i>=0;i--){
            b[i]=a[i]+b[i+1];
        }
        int ans=1,m,j;
        for(int nh=0;nh<q;nh++){
            m=sc.nextInt();
            x=0;
            y=v;
            while(x<=y){
                int z=(x+y)/2;
                int i=upper_bound(a, z);
                if(i<n){
                    j=b[i];
                    j-=(n-i)*z;
                    if(j>=m){
                        x=z+1;
                        ans = z;
                    }
                    else{
                        y=z-1;
                    }
                }
                else{
                    y=z-1;
                }
            }
            System.out.print(ans+" ");
        }
        System.out.println();
        
    



    
    }
}
